prow/pluginhelp/hook/hook.go

```diff
@@ -0,0 +1,366 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package hook provides the plugin help components to be compiled into the hook binary.
+// This includes the code to fetch help from normal and external plugins and the code to build and
+// serve a pluginhelp.Help struct.
+package hook
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/pluginhelp/externalplugins"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+// TODO: unit test to ensure that external plugins with the same name have the same endpoint and events.
+
+const (
+	// newRepoDetectionLimit is the maximum allowable time before a repo will appear in the help
+	// information if it is a new repo that is only referenced via its parent org in the config.
+	// (i.e. max time before help is available for brand new "kubernetes/foo" repo if only
+	// "kubernetes" is listed in the config)
+	newRepoDetectionLimit = time.Hour
+)
+
+type pluginAgent interface {
+	Config() *plugins.Configuration
+}
+
+type githubClient interface {
+	GetRepos(org string, isUser bool) ([]github.Repo, error)
+}
+
+type HelpAgent struct {
+	log *logrus.Entry
+	pa  pluginAgent
+	oa  *orgAgent
+}
+
+func NewHelpAgent(pa pluginAgent, ghc githubClient) *HelpAgent {
+	l := logrus.WithField("client", "plugin-help")
+	return &HelpAgent{
+		log: l,
+		pa:  pa,
+		oa:  newOrgAgent(l, ghc, newRepoDetectionLimit),
+	}
+}
+
+func (ha *HelpAgent) generateNormalPluginHelp(config *plugins.Configuration, revMap map[string][]string) (allPlugins []string, pluginHelp map[string]pluginhelp.PluginHelp) {
+	pluginHelp = map[string]pluginhelp.PluginHelp{}
+	for name, provider := range plugins.HelpProviders() {
+		allPlugins = append(allPlugins, name)
+		if provider == nil {
+			ha.log.Warnf("No help is provided for plugin '%s'.", name)
```

Comment:
nit: you can use `%q` instead of `'%s'`

Reply:
I was trying to avoid escaping the `"`s in the final log output. For example: `"error":"error creating prow job: body: {\"kind\":\"Status\",...`. This isn't important though. I'll change it to `%q` for consistency.